### PR TITLE
Increase disterl buffer sizes.

### PR DESCRIPTION
**Pull request by evanmcc**
At this point in riak's history disterl is our primary bottleneck.
Something about the way that it is framing small messages is causing its
tcp connection to size its send and recv buffers much smaller than
is ideal.

This change adds code to automatically set dist port sizes on nodeup,
as well as code to manually set all disterl connections to specific
sizes, meant to be used from the console for testing and tuning.

### CODE REVIEW

**Comment by evanmcc** [comment]
distbufs-base-pre3base-uniform-fixed-write_heavy-binsize1024-2013-10-4-20-9-51vsdistbufs-pre3bufs-uniform-fixed-write_heavy-binsize1024-2013-10-4-18-55-38-summary
This is the perf run.  I am told that the baseline run (on the left) for develop may be especially slow because of a known regression.

**Comment by evanmcc** [comment]
distbufs-base142-1 4 2-uniform-fixed-write_heavy-binsize1024-2013-10-5-14-48-28vsdistbufs-pre3bufs-uniform-fixed-write_heavy-binsize1024-2013-10-4-18-55-38-summary

1.4.2 doesn't fare a ton better.

File: ebin/riak_core.app [hunk 1]
```diff
@@ -32,6 +32,8 @@
              riak_core_console,
              riak_core_coverage_fsm,
              riak_core_coverage_plan,
+             riak_core_dist_mon,
+             riak_core_dist_mon_sup,
```
Comment (slfritchie) [inline ebin/riak_core.app]: I'm curious about why add a supervisor.  Do you have plans to add additional worker processes?

Comment (evanmcc) [inline ebin/riak_core.app]: I didn't know whether or not I needed to. I know Erlang the language OK at this point, but the proper OTP idioms are still somewhat opaque to me.  Suggestions as to how to properly structure a piece of code like this are welcomed.

**Comment by slfritchie** [comment]
I pushed a couple of small refactoring patches, if you like.  They compile without errors.  :-)

**Comment by evanmcc** [comment]
updated

File: ebin/riak_core.app [hunk 2]
```diff
@@ -146,6 +147,10 @@
 
          %% Handoff IP/port
          {handoff_port, 8099},
-         {handoff_ip, "0.0.0.0"}
+         {handoff_ip, "0.0.0.0"},
+         
+         %% Disterl buffer sizes
+         {dist_send_buf_size, 393216},
```
Comment (Vagabond) [inline ebin/riak_core.app]: What unit are these in and why were these particular numbers chosen? Can we record that for posterity here in this file?

Comment (evanmcc) [inline ebin/riak_core.app]: bytes, scattershot empirical testing/what worked on my machines.  More science could likely be applied.  I'll add some comments.

File: src/riak_core_sup.erl [hunk 3]
```diff
@@ -46,10 +48,14 @@ start_link() ->
 %% ===================================================================
 
 init([]) ->
+    DistMonEnabled = app_helper:get_env(riak_core, disable_dist_mon,
+                                        false) == false,
     Children = lists:flatten(
                  [?CHILD(riak_core_sysmon_minder, worker),
                   ?CHILD(riak_core_vnode_sup, supervisor, 305000),
                   ?CHILD(riak_core_eventhandler_sup, supervisor),
+                  ?IF(DistMonEnabled,
```
Comment (Vagabond) [inline src/riak_core_sup.erl]: You could do [?CHILD(riak_core_dist_mon, worker) || DistMonEnabled] instead of making a macro....

**Comment by Vagabond** [comment]
Forgive my ignorance, but how does this interact with the zddbl flag, if at all?

File: src/riak_core_sup.erl (hunk 3 above)
Comment (evanmcc) [inline src/riak_core_sup.erl]: I pulled the macro from riak_kv_sup.  I didn't realize that list comprehensions could be bent that way.

**Comment by evanmcc** [comment]
My understanding is that they're entirely separate.  I'll look into it.

**Comment by Vagabond** [comment]
I does seem to be independant, yes.

File: src/riak_core_sup.erl [hunk 4]
```diff
@@ -46,10 +48,14 @@ start_link() ->
 %% ===================================================================
 
 init([]) ->
+    DistMonEnabled = app_helper:get_env(riak_core, disable_dist_mon,
```
Comment (Vagabond) [inline src/riak_core_sup.erl]: Backwards flags like this are hard to understand, why not use enable_dist_mon?

**Comment by Vagabond** [comment]
I'm seeing some odd results in theshell:

```
(dev1@127.0.0.1)13> net_kernel:monitor_nodes(true, [{node_type, visible}, nodedown_reason]).
ok
(dev1@127.0.0.1)14> nodes().
[]
(dev1@127.0.0.1)15> flush()
Shell got {nodedown,'dev2@127.0.0.1',
                    [{nodedown_reason,connection_closed},{node_type,visible}]}
Shell got {nodeup,'dev2@127.0.0.1',[{node_type,visible}]}
ok
(dev1@127.0.0.1)16> nodes().
['dev2@127.0.0.1']
(dev1@127.0.0.1)17> [{Name, inet:getopts(Port, [sndbuf, recbuf])} || {Name, Port} <- erlang:system_info(dist_ctrl)].
[{'dev2@127.0.0.1',{ok,[{sndbuf,425984},{recbuf,425984}]}},
 {'c_29911_dev1@127.0.0.1',{ok,[{sndbuf,663750},
                                {recbuf,262030}]}}]
(dev1@127.0.0.1)18> riak_core_dist_mon:get_riak_env_vars().
{393216,786432}
```

**Comment by Vagabond** [comment]
Oh, nevermind. net.core.[rw]mem_max was screwing me. Once I bumped that to 6mb the values reported were esactly double what was set in the config file.

Other than the previous comments, this seems to work fine.

+1

File: src/riak_core_sup.erl (hunk 4 above)
Comment (evanmcc) [inline src/riak_core_sup.erl]: My thinking was that since it's enabled by default, one would need to disable it, but it does look really ugly.

**Pull request merged**